src/main/java/com/google/sps/data/Election.java

```diff
@@ -14,81 +14,166 @@
+   *
+   * @param electionQueryData the JSON output of an electionQuery to the Google Civic Information
+   *     API
+   * @return the new Election object
+   */
+  public static Election fromElectionQuery(JSONObject electionQueryData) throws JSONException {
     return Election.builder()
-        .setId(obj.getLong("id"))
-        .setName(obj.getString("name"))
-        .setDate(obj.getString("electionDay"))
-        .setScope(obj.getString("ocdDivisionId"))
-        // TODO(caseyprice): get values for Contests and Propositions
-        .setContests(new HashSet<Long>())
-        .setPropositions(new HashSet<Long>())
+        .setId(electionQueryData.getString("id"))
+        .setName(electionQueryData.getString("name"))
+        .setDate(electionQueryData.getString("electionDay"))
+        .setScope(electionQueryData.getString("ocdDivisionId"))
+        .setContests(new HashSet<String>())
+        .setPropositions(new HashSet<String>())
         .build();
   }
 
-  // creates a new Entity and sets the proper properties.
-  public Entity toEntity() {
+  /**
+   * Creates a new Election object by extracting the properties from an Election object and from the
+   * properties of "voterInfoQueryData". Delegates creating Contest Entities from this election's
+   * list of contests.
+   *
+   * @param election the Election object already populated by fromElectionQuery
+   * @param datastore the Datastore to store this election's list of contests
+   * @param voterInfoQueryData the JSON output of a voterInfoQuery to the Google Civic Information
+   *     API
+   * @return the new Election object
+   */
+  public static Election fromVoterInfoQuery(
+      Election election, DatastoreService datastore, JSONObject voterInfoQueryData)
+      throws JSONException {
+    HashSet<String> contestKeyList = new HashSet<String>();
+    HashSet<String> propositionKeyList = new HashSet<String>();
+
+    if (voterInfoQueryData.has("contests")) {
+      JSONArray contestListData = voterInfoQueryData.getJSONArray("contests");
+      for (Object contestObject : contestListData) {
+        JSONObject contest = (JSONObject) contestObject;
+
+        Contest.fromVoterInfoQuery(datastore, contest).putInDatastore(datastore);
+      }
+
+      contestKeyList = ServletUtils.getEntityKeyNameList(datastore, "Contest");
+    }
+
+    // TODO(caseyprice): get values for propositions
+
+    return Election.builder()
+        .setId(election.getId())
+        .setName(election.getName())
+        .setDate(election.getDate())
+        .setScope(election.getScope())
+        .setContests(contestKeyList)
+        .setPropositions(propositionKeyList)
+        .build();
+  }
+
+  /**
+   * Checks if an Election object has been populated by the output of a voterInfoQuery call from the
+   * Google Civic Information API.
+   *
+   * @return true if contests and propositions contain elements, false otherwise
+   */
+  public boolean isPopulatedByVoterInfoQuery() {
```

Comment:
Is an Election object that has empty contest / proposition data considered to be in an invalid state?
If so, is it possible to delay creating an Election object until we have the contest and proposition data also?

Generally it's best to avoid inserting "invalid" or incomplete data into the datastore as much as we can.  Sometimes this isn't possible though, but it makes handling weird asynchronous cases a lot easier.  For example it's probably possible that the website could read the Election data after it's been inserted into Datastore, but before it has its contest data populated.

Reply:
Yeah, we use Datastore Election entities without contests/propositions filled out for listing elections on the electionlist.html page. It's not necessarily invalid I think, just incomplete. In this case, we check if it's been populated only to avoid calling the API again when it's not necessary.